lightning/src/onion_message/blinded_route.rs

```diff
@@ -0,0 +1,160 @@
+	/// will be the destination node.
+	///
+	/// Errors if less than two hops are provided or if `node_pk`(s) are invalid.
+	pub fn new<Signer: Sign, K: KeysInterface, T: secp256k1::Signing + secp256k1::Verification>
+		(mut node_pks: Vec<PublicKey>, keys_manager: &K, secp_ctx: &Secp256k1<T>) -> Result<Self, ()>
+	{
+		if node_pks.len() < 2 { return Err(()) }
+		let blinding_secret_bytes = keys_manager.get_secure_random_bytes();
+		let blinding_secret = SecretKey::from_slice(&blinding_secret_bytes[..]).expect("RNG is busted");
+		let mut blinded_node_ids_and_payload_keys = construct_blinded_route_keys(secp_ctx, &node_pks, &blinding_secret).map_err(|_| ())?;
+		let mut route_keys = blinded_node_ids_and_payload_keys.drain(..);
+		let mut blinded_hops = Vec::with_capacity(node_pks.len());
+
+		let mut iter = node_pks.drain(..);
+		let introduction_node_id = iter.next().unwrap();
+		for pk in iter {
+			let (encrypted_payload_key, blinded_node_id) = route_keys.next().unwrap();
+			let payload = ForwardTlvs {
+				next_node_id: pk.clone(),
+				next_blinding_override: None,
+			};
+			blinded_hops.push(BlindedHop {
+				blinded_node_id,
+				encrypted_payload: encrypt_payload(payload, encrypted_payload_key),
+			});
+		}
+
+		// Add the recipient final payload.
+		let payload = ReceiveTlvs { path_id: None };
+		let (encrypted_payload_key, blinded_node_id) = route_keys.next().unwrap();
+		blinded_hops.push(BlindedHop {
+			blinded_node_id,
+			encrypted_payload: encrypt_payload(payload, encrypted_payload_key),
+		});
+
+		Ok(BlindedRoute {
+			introduction_node_id,
+			blinding_point: PublicKey::from_secret_key(secp_ctx, &blinding_secret),
+			blinded_hops,
+		})
+	}
+
+}
+
+/// Construct keys for creating a blinded route along the given `unblinded_path`.
+///
+/// Returns: `Vec<(encrypted_payload_key, blinded_node_id)>`
+/// where the former is for encrypting [`BlindedHop::encrypted_payload`] and the latter for
+/// [`BlindedHop::blinded_node_id`].
+fn construct_blinded_route_keys<T: secp256k1::Signing + secp256k1::Verification>(
+	secp_ctx: &Secp256k1<T>, unblinded_path: &Vec<PublicKey>, session_priv: &SecretKey
+) -> Result<Vec<([u8; 32], PublicKey)>, secp256k1::Error> {
+	let mut route_keys = Vec::with_capacity(unblinded_path.len());
+
+	utils::construct_keys_callback(secp_ctx, unblinded_path, None, session_priv, |blinded_hop_pubkey, _, _, _, encrypted_payload_ss| {
+		route_keys.push((encrypted_payload_ss, blinded_hop_pubkey));
+	})?;
+
+	Ok(route_keys)
+}
+
+/// Encrypt TLV payload to be used as a [`BlindedHop::encrypted_payload`].
+fn encrypt_payload<P: Writeable>(payload: P, encrypted_tlvs_ss: [u8; 32]) -> Vec<u8> {
+	let mut writer = VecWriter(Vec::new());
+	let write_adapter = ChaChaPolyWriteAdapter::new(encrypted_tlvs_ss, &payload);
+	write_adapter.write(&mut writer).expect("In-memory writes cannot fail");
+	writer.0
+}
+
+/// TLVs to encode in an intermediate onion message packet's hop data. When provided in a blinded
```

Comment:
I'm confused how it could only mean the last hop in the unblinded path? I updated the docs in bae356e361ccd3373722eac85f4058ff31dc3e53, let me know if there's some other way to clarify them

Reply:
Yeah that's why I understand from the spec wording here :

"They communicate this blinded route and the encrypted blobs to the sender. The sender finds a route to the introduction node and extends it with the blinded route provided by the recipient" Though not big concern if we differ.